src/Airship/Internal/Helpers.hs

```diff
@@ -94,21 +97,74 @@ toWaiResponse Response{..} cfg trace quip =
+    resourceToWaiT cfg (const id) routes errors
+
+-- | Given a 'RoutingSpec', an 'ErrorResponses', and a user state @s@, construct a WAI 'Application'.
+resourceToWaiT :: Monad m =>
+                  AirshipConfig
+               -> (Request -> m Wai.Response -> IO Wai.Response)
+               -> RoutingSpec m ()
+               -> ErrorResponses m
+               -> Wai.Application
+resourceToWaiT cfg run routes errors req respond = do
     let routeMapping = runRouter routes
         pInfo = Wai.pathInfo req
-        (resource, (params', matched)) = route routeMapping pInfo resource404
-    nowTime <- getCurrentTime
     quip <- getQuip
-    (=<<) respond . run req $ do
-      (response, trace) <- eitherResponse nowTime params' matched req (flow resource)
-      return $ toWaiResponse response cfg (traceHeader trace) quip
+    nowTime <- getCurrentTime
+    case route routeMapping pInfo of
+        Nothing -> do
+            -- No route matched so bail out
+            let default404Resp = Response HTTP.status404 [(HTTP.hContentType, "text/plain")] Empty
+            respond =<< run req (do
+                (response, trace) <- eitherResponse nowTime mempty [] req (errorResponse errors default404Resp)
+                return $ toWaiResponse response cfg (traceHeader trace) quip)
+        Just (resource, (params', matched)) -> do
+            -- Merge the application-wide error responses with the resource
+            -- specific error responses favoring the resource-specific
+            --in case of conflict
+            let mergedResource = resource { errorResponses = M.union (errorResponses resource) errors }
+            respond =<< run req (do
+                (response, trace) <- eitherResponse nowTime params' matched req (flow mergedResource >>= errorResponse (errorResponses mergedResource))
+                return $ toWaiResponse response cfg (traceHeader trace) quip)
+
+
+-- | If the Response body is Empty the response body is set based on the error responses
+-- | provided by the application and resource. If the response body is not Empty or
+-- | there are no error response configured for the status code in the Response then no
+-- | action is taken. The contents of the 'Webmachine'' response body will be streamed
+-- | back to the client.
+errorResponse :: Monad m =>
+                 ErrorResponses m
+              -> Response
+              -> Webmachine m Response
+errorResponse errResps r@Response{..}
+    | (HTTP.statusIsClientError _responseStatus ||
+       HTTP.statusIsServerError _responseStatus) &&
+       isResponseBodyEmpty _responseBody = do
+           req <- request
+           let reqHeaders = requestHeaders req
+               acceptStr = lookup HTTP.hAccept reqHeaders
+               errBodies = map dupContentType <$> M.lookup _responseStatus errResps
+               mResp = join $ mapAcceptMedia <$> errBodies <*> acceptStr
+           when (isJust mResp) $ do
+               let (Just (ct, body)) = mResp
```

Comment:
Just a nit reminder, but would definitely prefer to use `forM` here rather than having to use an unsafe cast.